Corona-Warn-App/src/main/java/de/rki/coronawarnapp/coronatest/TestCertificateRepository.kt

```diff
@@ -0,0 +1,329 @@
+import de.rki.coronawarnapp.covidcertificate.test.TestCertificateQRCodeExtractor
+import de.rki.coronawarnapp.util.TimeStamper
+import de.rki.coronawarnapp.util.coroutine.AppScope
+import de.rki.coronawarnapp.util.coroutine.DispatcherProvider
+import de.rki.coronawarnapp.util.encryption.rsa.RSACryptography
+import de.rki.coronawarnapp.util.encryption.rsa.RSAKeyPairGenerator
+import de.rki.coronawarnapp.util.flow.HotDataFlow
+import de.rki.coronawarnapp.util.mutate
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.SharingStarted
+import kotlinx.coroutines.flow.catch
+import kotlinx.coroutines.flow.first
+import kotlinx.coroutines.flow.launchIn
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.flow.onEach
+import kotlinx.coroutines.flow.onStart
+import kotlinx.coroutines.plus
+import kotlinx.coroutines.withContext
+import okio.ByteString.Companion.decodeBase64
+import org.joda.time.Duration
+import timber.log.Timber
+import java.util.UUID
+import javax.inject.Inject
+import javax.inject.Singleton
+
+@Singleton
+class TestCertificateRepository @Inject constructor(
+    @AppScope private val appScope: CoroutineScope,
+    private val dispatcherProvider: DispatcherProvider,
+    private val timeStamper: TimeStamper,
+    private val storage: TestCertificateStorage,
+    private val certificateServer: CovidCertificateServer,
+    private val rsaKeyPairGenerator: RSAKeyPairGenerator,
+    private val rsaCryptography: RSACryptography,
+    private val qrCodeExtractor: TestCertificateQRCodeExtractor,
+    private val appConfigProvider: AppConfigProvider,
+) {
+
+    private val internalData: HotDataFlow<Map<TestCertificateIdentifier, TestCertificateContainer>> = HotDataFlow(
+        loggingTag = TAG,
+        scope = appScope + dispatcherProvider.Default,
+        sharingBehavior = SharingStarted.Eagerly,
+    ) {
+        storage.testCertificates.map { it.identifier to it }.toMap().also {
+            Timber.tag(TAG).v("Restored TestCertificate data: %s", it)
+        }
+    }
+
+    val certificates: Flow<Set<TestCertificateContainer>> = internalData.data.map { it.values.toSet() }
+
+    init {
+        internalData.data
+            .onStart { Timber.tag(TAG).d("Observing TestCertificateContainer data.") }
+            .onEach {
+                Timber.tag(TAG).v("TestCertificateContainer data changed: %s", it)
+                storage.testCertificates = it.values.toSet()
+            }
+            .catch {
+                it.reportProblem(TAG, "Failed to snapshot TestCertificateContainer data to storage.")
+                throw it
+            }
+            .launchIn(appScope + dispatcherProvider.Default)
+    }
+
+    suspend fun requestCertificate(test: CoronaTest): TestCertificateContainer {
+        Timber.tag(TAG).d("createDccForTest(test.identifier=%s)", test.identifier)
```

Comment:
For an existing corona test, create a new test certificate entry. This just creates the entry to kick off automation.
No network requests happen yet.